Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/repo/DefaultContactDiaryRepository.kt

```diff
@@ -0,0 +1,205 @@
+import de.rki.coronawarnapp.contactdiary.storage.entity.toContactDiaryPersonEntity
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.map
+import org.joda.time.LocalDate
+import timber.log.Timber
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class DefaultContactDiaryRepository @Inject constructor(
+    private val contactDiaryLocationDao: ContactDiaryLocationDao,
+    private val contactDiaryLocationVisitDao: ContactDiaryLocationVisitDao,
+    private val contactDiaryPersonDao: ContactDiaryPersonDao,
+    private val contactDiaryPersonEncounterDao: ContactDiaryPersonEncounterDao
+) : ContactDiaryRepository {
+
+    // Location
+    override val locations: Flow<List<ContactDiaryLocation>> = contactDiaryLocationDao
+        .allEntries()
+        .map { it.sortByNameAndIdASC() }
+
+    override suspend fun addLocation(contactDiaryLocation: ContactDiaryLocation) {
+        Timber.d("Adding location $contactDiaryLocation")
+        contactDiaryLocationDao.insert(contactDiaryLocation.toContactDiaryLocationEntity())
+    }
+
+    override suspend fun updateLocation(contactDiaryLocation: ContactDiaryLocation) {
+        Timber.d("Updating location $contactDiaryLocation")
+        val contactDiaryContactDiaryLocationEntity = contactDiaryLocation.toContactDiaryLocationEntity()
+        executeWhenIdNotDefault(contactDiaryContactDiaryLocationEntity.locationId) {
+            contactDiaryLocationDao.insert(contactDiaryContactDiaryLocationEntity)
+        }
+    }
+
+    override suspend fun deleteLocation(contactDiaryLocation: ContactDiaryLocation) {
+        Timber.d("Deleting location $contactDiaryLocation")
+        val contactDiaryContactDiaryLocationEntity = contactDiaryLocation.toContactDiaryLocationEntity()
+        executeWhenIdNotDefault(contactDiaryContactDiaryLocationEntity.locationId) {
+            contactDiaryLocationDao.delete(contactDiaryContactDiaryLocationEntity)
+        }
+    }
+
+    override suspend fun deleteLocations(contactDiaryLocations: List<ContactDiaryLocation>) {
+        Timber.d("Deleting location $contactDiaryLocations")
+        val contactDiaryLocationEntities = contactDiaryLocations
+            .map {
+                val contactDiaryLocationEntity = it.toContactDiaryLocationEntity()
+                executeWhenIdNotDefault(contactDiaryLocationEntity.locationId)
+                return@map contactDiaryLocationEntity
+            }
+        contactDiaryLocationDao.delete(contactDiaryLocationEntities)
+    }
+
+    override suspend fun deleteAllLocations() {
+        Timber.d("Clearing contact diary location table")
+        contactDiaryLocationDao.deleteAll()
+    }
+
+    // Location visit
+    private suspend fun List<ContactDiaryLocationVisitEntity>.toContactDiaryLocationVisitSortedList():
+        List<ContactDiaryLocationVisit> =
+        this.map {
+            val contactLocation = contactDiaryLocationDao.entityForId(id = it.fkLocationId)
```

Comment:
Why do you join the data on app-level and not in the DB? For `n` entities, you would have `n+1` SQL statements, while you could have just one...

Reply:
I will try to change this...

Reply:
adjusted.
https://github.com/corona-warn-app/cwa-app-android/pull/1870/commits/93b39f0daf809049313dd036f69b8f885aa34d78